docs/rfc/009-bit-byte-conversions.md

```diff
@@ -8,237 +8,330 @@ The Aleo Team.
-it makes sense to generate them programmatically (e.g. via Rust macros),
-rather than enumerating all of them explicitly in the implementation.
-It may also makes sense, at R1CS generation time,
-to use generated or suitably parameterized code to recognize them and turn them into the corresponding gadgets.
+### Conversions with Group Elements
+
+A group element consists of two field elements that satisfy the elliptic curve equation,
+i.e. a group element is a point on the curve on the plane defined by the cartesian product of the field with itself.
+The point consists of an _x_ and a _y_ coordinate.
+
+A group element is converted to a sequence of `N` bits,
+where `N` is twice the bit length of the prime number that defines the field,
+by juxtaposing the `N/2` bits obtained from converting the _x_ and _y_ coordinates as field elements.
+More precisely:
+* The conversion of a group element (_x_, _y_) to big endian bits is `[...x, ...y]`,
+  where `x` is the conversion of _x_ to big endian bits and `y` is the conversion of _y_ to big endian bits.
+* The conversion of a group element (_x_, _y_) to little endian bits is `[...y, ...x]`,
+  where `x` is the conversion of _x_ to little endian bits and `y` is the conversion of _y_ to little endian bits.
+
+Reversing the order of the coordinates for little endian means that
+the little endian conversion of a group element is the reverse of the big endian conversion of the same group element,
+a property shared with the integer and field element conversions.
+However, given that a group element is not a single number,
+the notion of big and little endian as such does not directly apply to group elements,
+but only to their individual coordinates.
+
+Not all (little or big endian) sequences of `N` bits represent group elements.
+Not only each `N/2`-bit half must represent a field element (see discussion for field elements above),
+but also the resulting point (_x_, _y_) must satisfy the curve equation.
+Attempting to convert to a group element a sequence of `N` bits that does not actually represent a group element
+causes an error, in the same sense as division by zero causes an error.
+
+Conversions with byte sequences are handled by juxtaposition of the byte sequence representations of _x_ and _y_:
+* The conversion of a group element (_x_, _y_) to big endian bytes is `[...x, ...y]`,
+  where `x` is the conversion of _x_ to big endian bytes and `y` is the conversion of _y_ to big endian bytes.
+* The conversion of a group element (_x_, _y_) to little endian bytes is `[...y, ...x]`,
+  where `x` is the conversion of _x_ to little endian bytes and `y` is the conversion of _y_ to little endian bytes.
+
+For the currently supported elliptic curve in Leo, `N` is 506, and `M` is 64.
+These valuee will change with the curve, when Leo is independently extended with support for more curves,
+as discussed earlier for field elements.
+
+### Conversions with Characters
```

Comment:
This is the section that seems least intuitive.  Why can't the Leo compiler add the extra 0 bits or bytes when going from a valid 21-bool array to a char?  Why can't conversion from a char ignore the high-order bits or bytes and generate a 21-bool array?

Reply:
It should be possible; I think that @gluax proposed something along these lines. We would still pay for the extra R1CS constraints if we use the existing field gadgets though. I think the current plan was to try and develop a new gadget.